### ISSUE TITLE
AltCovCobertura: startIndex cannot be larger than length of string.

### ISSUE THREAD

**Issue by kMutagene** [SteveGilham/altcover#122]
Hi there, reporting a problem with AltCover.Cobertura.

I am using the following dotnet test setup (via this fake target):

```fsharp
let runTestsWithCodeCov = BuildTask.create "RunTestsWithCodeCov" [clean; build; copyBinaries] {
    let standardParams = Fake.DotNet.MSBuild.CliArguments.Create ()
    Fake.DotNet.DotNet.test(fun testParams ->
        {
            testParams with
                MSBuildParams = {
                    standardParams with
                        Properties = [
                            "AltCover","true"
                            "AltCoverCobertura","../../codeCov.xml"
                            "AltCoverForce","true"
                        ]
                };
                Logger = Some "console;verbosity=detailed"
        }
    ) testProject
}
```

so basically i am running `dotnet test testProject.fsproj /p:AltCover=true /p:AltCoverLcovReport=../../codeCov.txt /p:AltCoverForce=true`

you can see the error in CI on this PR: https://github.com/fslaborg/FSharp.Stats/pull/102

Which worked fine for a long time now, but now i get this error (log at the end of the issue). I for now changed the output format to lcov, which works just fine, so this seems to be a problem with cobertura output.

```
System.ArgumentOutOfRangeException: startIndex cannot be larger than length of string. (Parameter 'startIndex')
   at System.String.Substring(Int32 startIndex, Int32 length)
   at System.String.Substring(Int32 startIndex)
   at AltCover.Cobertura.I.arrangeMethods@224-2.Invoke(XElement method) in C:\projects\altcover\AltCover.Engine\Cobertura.fs:line 228
   at Microsoft.FSharp.Collections.Internal.IEnumerator.map@75.DoMoveNext(b& curr) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\seq.fs:line 81
   at Microsoft.FSharp.Collections.Internal.IEnumerator.MapEnumerator`1.System.Collections.IEnumerator.MoveNext() in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\seq.fs:line 68
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at Microsoft.FSharp.Collections.SeqModule.ToArray[T](IEnumerable`1 source)
   at Microsoft.FSharp.Collections.SeqModule.SortBy@1106.Invoke(Unit unitVar0) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\seq.fs:line 1107
   at Microsoft.FSharp.Collections.SeqModule.mkDelayedSeq@471.Invoke(Unit unitVar0) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\seq.fs:line 471
   at Microsoft.FSharp.Collections.SeqModule.revamp@576.Invoke(Unit unitVar0) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\seq.fs:line 576
   at Microsoft.FSharp.Collections.SeqModule.Fold[T,TState](FSharpFunc`2 folder, TState state, IEnumerable`1 source) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\seq.fs:line 727
   at AltCover.Cobertura.I.processClass@238-3.Invoke(Tuple`2 tupledArg, Tuple`2 tupledArg) in C:\projects\altcover\AltCover.Engine\Cobertura.fs:line 244
   at Microsoft.FSharp.Collections.SeqModule.Fold[T,TState](FSharpFunc`2 folder, TState state, IEnumerable`1 source) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\seq.fs:line 731
   at AltCover.Cobertura.I.openCover@274.Invoke(XElement module) in C:\projects\altcover\AltCover.Engine\Cobertura.fs:line 286
   at Microsoft.FSharp.Collections.SeqModule.Iterate[T](FSharpFunc`2 action, IEnumerable`1 source) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\seq.fs:line 497
   at AltCover.Cobertura.I.openCover(XDocument report, XElement packages) in C:\projects\altcover\AltCover.Engine\Cobertura.fs:line 288
   at AltCover.Cobertura.convertReport(XDocument report, ReportFormat format) in C:\projects\altcover\AltCover.Engine\Cobertura.fs:line 315
   at AltCover.Cobertura.summary[a](XDocument report, ReportFormat format, a result) in C:\projects\altcover\AltCover.Engine\Cobertura.fs:line 336
   at AltCover.Runner.declareOptions@579-10.Invoke(XDocument report, ReportFormat format, Int32 result) in C:\projects\altcover\AltCover.Engine\Runner.fs:line 579
   at AltCover.Runner.J.doSummaries@898-1.Invoke(FSharpFunc`2 summary) in C:\projects\altcover\AltCover.Engine\Runner.fs:line 898
   at Microsoft.FSharp.Collections.ListModule.Fold[T,TState](FSharpFunc`2 folder, TState state, FSharpList`1 list) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\list.fs:line 222
   at AltCover.Runner.J.doSummaries(XDocument document, ReportFormat format, Int32 result) in C:\projects\altcover\AltCover.Engine\Runner.fs:line 897
   at AltCover.Runner.value@929-3.Invoke(Unit unitVar0) in C:\projects\altcover\AltCover.Engine\Runner.fs:line 956
   at AltCover.CommandLine.I.doPathOperation[a](FSharpFunc`2 f, a defaultValue, Boolean store) in C:\projects\altcover\AltCover.Engine\CommandLine.fs:line 206
ActualValue = 
<null>
ParamName = 
"startIndex"
TargetSite = 
System.String Substring(Int32, Int32)
Data = 
seq []
InnerException = 
<null>
HelpLink = 
<null>
Source = 
"System.Private.CoreLib"
HResult = 
-2146233086
```

**Comment by SteveGilham** [comment]
Sorry about delay in replying here -- a lot of plates to keep spinning just ATM.

The problem is happening when trying to take a full method name `returntype class::method(arguments)` down to the cobertura style  `returntype method(arguments)`; at least as this is an open source project, I shouldn't have to worry about getting a repro case.

**Comment by SteveGilham** [comment]
The repro was immediate.  The problem was in the tokenization of the method name `TKey <;StartupCode$FSharp-Core>.$Map/System-Collections-Generic-IDictionary<'Key, 'Value>-get_Keys@669-1::Invoke(System.Collections.Generic.KeyValuePair`2<TKey,TValue>)` which has a space after the comma.  It's also shown up an off-by-one error in putting together the cobertura style method name and `returntype (arguments)` signature.

Now I need to make sure that the NCover to Cobertura transformation is also safe before calling this ready to add to the next release.

**Comment by SteveGilham** [comment]
For the moment. presuming that coverage of the system `FSharp.Core.dll` assembly is not of interest, adding the line 
```
                            "AltCoverAssemblyExcludeFilter","FSharp\.Core"
```
to the list of properties will exclude that assembly and thus the method that is causing this issue.  I've not yet tested that there no such method signatures in the `FSharp.Stats` code, though.

**Comment by kMutagene** [comment]
> Sorry about delay in replying here -- a lot of plates to keep spinning just ATM.

No worries, and in my book you actually looked really fast into the problem, cheers! 

I will try excluding FSharp.Core and report back.

**Comment by kMutagene** [comment]
Adding the FSharp.Core assembly to AltCover's exclude filter works for FSharp.Stats as well. Thank you for the quick response 👍 As lcov format works for me just fine, ill stick with it for a few builds, and if problems arise from that, i now have a backup plan, thanks again!

**Comment by SteveGilham** [comment]
Just ran the test myself -- and not only did it run through, it also dropped the amount of code being instrumented to ~40% of the "do everything" level, so filtering will be saving cycles in the build as well.

**Comment by SteveGilham** [comment]
This should be resolved in the new 8.0.815 release.
